gratipay/models/participant/email.py

```diff
@@ -41,94 +41,172 @@ class Email(object):
-                c.run("""
-                    INSERT INTO emails
-                                (address, nonce, verification_start, participant_id)
-                         VALUES (%s, %s, %s, %s)
-                """, (email, nonce, verification_start, self.id))
-        except IntegrityError:
-            nonce = self.db.one("""
+        if existing is None:
+
+            # Not in the table yet. This should throw an IntegrityError if the
+            # address is verified for a different participant.
+
+            c.run( "INSERT INTO emails (participant_id, address, nonce) VALUES (%s, %s, %s)"
+                 , (self.id, email, nonce)
+                  )
+        else:
+
+            # Already in the table. Restart verification. Henceforth, old links
+            # will fail.
+
+            if existing.nonce:
+                c.run('DELETE FROM claims WHERE nonce=%s', (existing.nonce,))
+            c.run("""
                 UPDATE emails
-                   SET verification_start=%s
+                   SET nonce=%s
+                     , verification_start=now()
                  WHERE participant_id=%s
                    AND address=%s
-                   AND verified IS NULL
-             RETURNING nonce
-            """, (verification_start, self.id, email))
-            if not nonce:
-                return self.add_email(email)
-
-        base_url = gratipay.base_url
-        username = self.username_lower
-        encoded_email = encode_for_querystring(email)
-        link = "{base_url}/~{username}/emails/verify.html?email2={encoded_email}&nonce={nonce}"
-        self.app.email_queue.put( self
-                                , 'verification'
-                                , email=email
-                                , link=link.format(**locals())
-                                , include_unsubscribe=False
-                                 )
-        if self.email_address:
-            self.app.email_queue.put( self
-                                    , 'verification_notice'
-                                    , new_email=email
-                                    , include_unsubscribe=False
+            """, (nonce, self.id, email))
 
-                                    # Don't count this one against their sending quota.
-                                    # It's going to their own verified address, anyway.
-                                    , _user_initiated=False
-                                     )
+        return nonce
+
+
+    def start_package_claims(self, c, nonce, *packages):
+        """Takes a cursor, nonce and list of packages, inserts into ``claims``
+        and returns ``None``.
+        """
+        if not packages:
+            return
+        VALUES, values = [], []
+        for p in packages:
+            VALUES.append('(%s, %s)')
+            values += [nonce, p.id]
+        VALUES = ', '.join(VALUES)
```

Comment:
It seems tricky to switch types of a variable from line to line. Maybe that's the Java coder in me though. ;-)